Add a callback for each retry

doits opened the pull request:
If you want something to run each time a job is retried, you can include a callback:

```ruby
include ActiveJob::Retry.new(
  callback: proc do |exception, delay|
    # will be run before queuing next retry
  end
)
```

Additionally, the callback can stop the retries by returning `:stop_retry`:

```ruby
include ActiveJob::Retry.new(
  callback: proc do |exception, delay|
    if some_condition
      :stop_retry
    end
  end
)
```

I can write some specs if you seem this a useful addition to merge in.

isaacseymour commented on lib/active_job/retry.rb:
```diff
@@ -32,9 +32,10 @@ class Retry < Module
+    def initialize(strategy: nil, callback: nil, **options)
       check_adapter!
       @backoff_strategy = choose_strategy(strategy, options)
+      @retry_callback = callback if callback.is_a?(Proc)
```
This should error if something other than a proc (or nil) is passed here.

isaacseymour commented:
I was thinking of adding something like this myself (for metric collection), so definitely keen for this!

doits commented:
added a validation and some specs <sup>(but I must admin the specs look a little bit messy)</sup>

the one spec error (something with the logger) is not from my change afaik, since it happens on master branch, too

doits commented:
... and added one more spec.

I can squash this all into one commit when you say you are fine with everything, so commit history does not get so messy.

isaacseymour commented on lib/active_job/retry.rb:
```diff
@@ -117,6 +125,18 @@ def check_adapter!
     end
 
+    def validate_params
+      if retry_callback && !retry_callback.is_a?(Proc)
```
Actually, maybe this should just be `retry_callback.responds_to?(:call) && retry_callback.method(:call).arity == 2`? Then it doesn't _have_ to be a proc - it can be anything "proc-like", which is probably more convenient.

doits replied:
It's a little bit more complicated since the current solution uses `instance_exec` to call the proc, so that `self` is the current instance. So I am not sure if the check is correct for this?

isaacseymour replied:
Ah, yeah good point. I wonder how valuable having it instance_exec'd would be - do you have an example usage where it's necessary?

doits replied:
Well, basically you can access everything you have defined for you current job instance, for example instance variables. Or methods of that instance (in particular the methods added by activejob-retry too, for example `serialize`). I currently use it like this:

    include ActiveJob::Retry.new(
      ...
      callback: proc do |exception, delay|
        Rails.logger.warn(
          "Retrying because of exception \"#{exception.message}\" in #{delay} seconds, retry attempt #{retry_attempt}"
        )
        ExceptionNotifier.notify_exception(
          exception, data: { delay: delay }.merge(serialize)
        )
      end
    )

isaacseymour commented on spec/retry_spec.rb:
```diff
@@ -295,4 +295,54 @@ def self.retry_delay(_attempt, _exception)
+    let(:retry_instance) do
+      described_class.new(strategy: :constant, callback: callback, **options)
+    end
+    let(:callback) { proc { raise 'callback executed' } }
```
(if the validation is just on responding to `call`, this could be an RSpec `double(call: nil)` here - and then the assertions can be on stuff like `expect(callback).to have_received(:call)`, which feels neater)

doits replied:
Same here, `#call` is never called by `instance_exec`. How to create a double that can be `instance_exec`ed?

isaacseymour replied:
This is also kinda weird, but could do:
```ruby
let(:callback_double) { double(call: nil) }
let(:callback) { callback_double.method(:call).to_proc }
```

isaacseymour commented:
Sorry for being so slow reviewing this. I've pushed a commit to master to get Rubocop back to being happy. If you can rebase, the tests should pass 💃

doits commented:
I changed the spec and rebased everything into one commit on top of master. Looks like ruby 2 is not supported by all dependencies (nokogiri), so it should be simply dropped from CI IMO. Since ruby 2.0 is not support officially anyway.

isaacseymour commented:
Agreed, 2.0 doesn't get security updates anymore. I've pushed another commit to master which I really hope passes! Another rebase and I'll merge. Sorry it's taken so long to get this merged, and thanks for the contribution 😄

doits commented:
the `:callback_double` thing is interesting, will keep that in mind.

I added some points to the readme and rebased it, so everything should be fine now.

isaacseymour commented:
Released in 0.6.3. Thanks again for the contribution 😄
